### PR TITLE
AbstractBoundInstrument refcounts and keeps the state of mapped.

### PR DESCRIPTION
**Pull request by bogdandrutu**
Split from #831

### CODE REVIEW

File: sdk/src/main/java/io/opentelemetry/sdk/metrics/AbstractBoundInstrument.java [hunk 1]
```diff
@@ -17,35 +17,64 @@
 package io.opentelemetry.sdk.metrics;
 
 import io.opentelemetry.metrics.InstrumentWithBinding.BoundInstrument;
-import io.opentelemetry.metrics.LabelSet;
+import java.util.concurrent.atomic.AtomicLong;
 
+/**
+ * Abstract class that extends the functionality of the BoundInstrument.
```
Comment (jkwatson) [inline sdk/src/main/java/io/opentelemetry/sdk/metrics/AbstractBoundInstrument.java]:
```suggestion
 * Abstract class that implements the basic functionality of the BoundInstrument.
```

File: sdk/src/main/java/io/opentelemetry/sdk/metrics/DoubleCounterSdk.java [hunk 2]
```diff
@@ -38,14 +38,14 @@ private DoubleCounterSdk(
...
+    BoundInstrument boundInstrument = bind(labelSet);
+    boundInstrument.add(delta);
+    boundInstrument.unbind();
   }
 
   @Override
-  public BoundDoubleCounter bind(LabelSet labelSet) {
-    return new BoundInstrument(labelSet, monotonic);
+  public BoundInstrument bind(LabelSet labelSet) {
+    return new BoundInstrument(monotonic);
```
Comment (jkwatson) [inline sdk/src/main/java/io/opentelemetry/sdk/metrics/DoubleCounterSdk.java]: why did you stop passing in the labelset?

File: sdk/src/main/java/io/opentelemetry/sdk/metrics/AbstractBoundInstrument.java (hunk 1 above)
Comment (bogdandrutu) [reply sdk/src/main/java/io/opentelemetry/sdk/metrics/AbstractBoundInstrument.java]: done.

File: sdk/src/main/java/io/opentelemetry/sdk/metrics/DoubleCounterSdk.java (hunk 2 above)
Comment (bogdandrutu) [reply sdk/src/main/java/io/opentelemetry/sdk/metrics/DoubleCounterSdk.java]: The mapping between LabelSet and bound will be in the AbstractInstrumentWithBinding (from #831 ) so unnecessary memory usage in the Bound to also keep a reference to the LabelSet.

File: sdk/src/main/java/io/opentelemetry/sdk/metrics/DoubleCounterSdk.java [hunk 3]
```diff
@@ -73,13 +73,12 @@ public int hashCode() {
...
-  private static final class BoundInstrument extends AbstractBoundInstrument
-      implements BoundDoubleCounter {
+  static final class BoundInstrument extends AbstractBoundInstrument implements BoundDoubleCounter {
 
     private final boolean monotonic;
 
-    BoundInstrument(LabelSet labelSet, boolean monotonic) {
-      super(labelSet);
+    BoundInstrument(boolean monotonic) {
+      super(null);
```
Comment (jkwatson) [inline sdk/src/main/java/io/opentelemetry/sdk/metrics/DoubleCounterSdk.java]: don't we need to be passing the labelset in to the ABI here?

Comment (bogdandrutu) [reply sdk/src/main/java/io/opentelemetry/sdk/metrics/DoubleCounterSdk.java]: This is now the aggregation, will pass in a later PR :)

**Pull request merged**